### ISSUE TITLE
Wrong pager with the API v8

### ISSUE THREAD

**Issue by sylvainvanhoof** [odoo/odoo#4944]
Hello, 

I discovered a new bug with the pager.
This problem appends with the latest version of Odoo.

### Demo data ###
You'll find a module to demonstrate the bug => https://owncloud.sylvainvh.be/index.php/s/zcj7B30VwoQ4DRA 
Don't forget to activate demo data.

### What is my problem ###
In my project, I use the new API (v8.0) and I've to redefine some methods. 
I redefined the method "search" with the decorator @api.model:
```python
@api.model
def search(self, args, offset=0, limit=None, order=None, count=False):
    result = super(ResPartner, self).search(args, offset=offset, limit=limit, order=order, count=count)
    return result
```
Now, the pager counter display only one element (see the screenshot https://owncloud.sylvainvh.be/index.php/s/gEMqdb5CBY0VVsB)

### Why ###
This problem originates from the switch between the old and the new API.
The number of pages in the view list comes from the method search_count()
```python
def search_count(self, cr, user, args, context=None):
    """ search_count(args) -> int

    Returns the number of records in the current model matching :ref:`the
    provided domain <reference/orm/domains>`.
    """
    res = self.search(cr, user, args, context=context, count=True)
    if isinstance(res, list):
        return len(res)
    return res
```
As you can see, this method calls the method search().
With the old API, the method search can return a list of ID OR the number of element (according the parameter count) and this is the problem. 

When the method search_count() call the method search, **the wrapper take the result to create a browse record set. The result will be a list with ONE record set and his ID will be the number of elements.**

### How to temporarily fix this ###
To fix this temporarily, I can remove the decorator. In this case, the wrapper will call this method with the old api
```python
def search(self, cr, uid, args, offset=0, limit=None, order=None, count=False, context=None):
    res = super(res_partner, self).search(cr, uid, args, offset=offset, limit=limit, order=order, count=count, context=context)
    return res
```

Thank for all.

Best regards,
Sylvain Van Hoof for Eezee-it

**Comment by rco-odoo** [comment]
@sylvainvanhoof As you noted by yourself, the bug is indeed caused by the decorator `@api.returns` which incorrectly wraps the result of the old-api call into a recordset.

This bug is a known one, and I have not found an easy way to fix it. The problem is the inconsistent signature of the method: it returns results of different types depending on some input parameter. Fixing this would almost require specific decorators for that method only!

We apologize for the annoyance it caused you. At the moment, your temporary fix is probably the best workaround so far.

**Comment by rco-odoo** [comment]
I will not close this issue now, since we don't have a proper solution yet.
I will simply label it as "wishlist", to keep it visible.

**Comment by rco-odoo** [comment]
A clean fix for this issue at #9724.
